{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import copy\n",
    "import json\n",
    "import re\n",
    "import json\n",
    "import pickle as pkl\n",
    "from importlib import reload\n",
    "from zipfile import ZipFile\n",
    "from io import BytesIO\n",
    "from collections import defaultdict\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from owlready2 import *\n",
    "import rdflib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "pc = get_ontology(\"../../data/PathwayCommons12.reactome.BIOPAX.owl\").load()\n",
    "#pc = get_ontology(\"../../data/PathwayCommons12.kegg.BIOPAX.owl\").load()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "Pathway = list(filter(lambda x: x.name == 'Pathway', pc.world.classes()))[0]\n",
    "Interaction = list(filter(lambda x: x.name == 'Interaction', pc.world.classes()))[0]\n",
    "Conversion = list(filter(lambda x: x.name == 'Conversion', pc.world.classes()))[0]\n",
    "Control = list(filter(lambda x: x.name == 'Control', pc.world.classes()))[0]\n",
    "Catalysis = list(filter(lambda x: x.name == 'Catalysis', pc.world.classes()))[0]\n",
    "TemplateReaction = list(filter(lambda x: x.name == 'TemplateReaction', pc.world.classes()))[0]\n",
    "BiochemicalReaction = list(filter(lambda x: x.name == 'BiochemicalReaction', pc.world.classes()))[0]\n",
    "BindingFeature = list(filter(lambda x: x.name == 'BindingFeature', pc.world.classes()))[0]\n",
    "ModificationFeature = list(filter(lambda x: x.name == 'ModificationFeature', pc.world.classes()))[0]\n",
    "Complex = list(filter(lambda x: x.name == 'Complex', pc.world.classes()))[0]\n",
    "\n",
    "#list(pc.world.classes())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_db_name(db):\n",
    "  if db == 'UniProtKB' or db == 'uniprot knowledgebase':\n",
    "    return 'uniprot'\n",
    "  elif db == 'ncbi gene':\n",
    "    return 'ncbi'\n",
    "  elif db == 'hgnc symbol':\n",
    "    return 'hgnc'\n",
    "  return db.lower()\n",
    "\n",
    "\n",
    "def convert_db_id(id):\n",
    "  if ':' in str(id):\n",
    "    return id.split(':')[-1].lower()\n",
    "  return str(id).lower()\n",
    "\n",
    "\n",
    "def make_db_ref(db, id):\n",
    "  return '{0}:{1}'.format(convert_db_name(db), convert_db_id(id))\n",
    "\n",
    "def load_gene_info(path):\n",
    "  \"\"\"Load the NCBI Gene file and returns a {gene_name: info} dictionary\"\"\"\n",
    "  gene_info = pd.read_csv(path, delimiter='\\t')\n",
    "  gene_info_map = {}\n",
    "  gene_name_map = {}\n",
    "  for _, row in gene_info.iterrows():\n",
    "    info = {\n",
    "      'name': row['Symbol'],\n",
    "#       'type': row['type_of_gene'],\n",
    "#       'desc': row['Other_designations'].split('|')[0],\n",
    "    }\n",
    "    ids = [make_db_ref('ncbi', row['GeneID'])]\n",
    "    gene_name_map[info['name']] = info\n",
    "    for xref in row['dbXrefs'].split('|'):\n",
    "      xref_parts = xref.split(':')\n",
    "      db = xref_parts[0]\n",
    "      id = xref_parts[-1]\n",
    "      ids.append(make_db_ref(db, id))\n",
    "    for id in ids:\n",
    "      gene_info_map[id] = info\n",
    "#     for s in row['Synonyms'].split('|'):\n",
    "#       gene_info_map[s] = info\n",
    "  return gene_info, gene_info_map, gene_name_map\n",
    "\n",
    "gene_info, gene_info_map, gene_name_map = load_gene_info('../../data/Mus_musculus.gene_info')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7425\n",
      "14\n",
      "12184\n",
      "12201\n",
      "19640\n"
     ]
    }
   ],
   "source": [
    "def parse_molecule(m):\n",
    "  res = {\n",
    "    'type': 'molecule',\n",
    "    'name': m.displayName if m.displayName is not None else m.name,\n",
    "#     'name': getattr(m, 'displayName', m.name),\n",
    "#     'name': m.name,\n",
    "  }\n",
    "  if isinstance(m, Complex):\n",
    "    res['type'] = 'complex'\n",
    "    res['component'] = [parse_molecule(cc) for cc in m.component]\n",
    "  if hasattr(m, 'cellularLocation') and m.cellularLocation is not None:\n",
    "    if len(m.cellularLocation.term) > 0:\n",
    "      res['cellularLocation'] = m.cellularLocation.term[0]\n",
    "      #print('cellularLocation', res['cellularLocation'])\n",
    "#   if hasattr(m, 'feature') and m.feature is not None:\n",
    "#     res['feature'] = []\n",
    "#     if len(m.feature) > 0:\n",
    "#       for f in m.feature:\n",
    "#         if not isinstance(f, ModificationFeature) and not isinstance(f, BindingFeature):\n",
    "#           continue\n",
    "#         feature = {\n",
    "#           'biopaxType': str(f.__class__).replace('biopax-level3.', ''),\n",
    "#         }\n",
    "#         if hasattr(f, 'modificationType') and f.modificationType is not None:\n",
    "#           feature['modificationType'] = f.modificationType.term[0]\n",
    "#         if hasattr(f, 'featureLocation') and f.featureLocation is not None:\n",
    "#           feature['featureLocation'] = [{\n",
    "#             'sequencePosition': l.sequencePosition,\n",
    "#             'positionStatus': l.positionStatus,\n",
    "#           } for l in f.featureLocation]\n",
    "# #         print(f.get_properties())\n",
    "# #         print(f.modificationType.term[0])\n",
    "# #         print(f.featureLocation[0].get_properties())\n",
    "# #         print(f.featureLocation[0].sequencePosition)\n",
    "# #         print(f.featureLocation[0].positionStatus)\n",
    "# # #         print(f.comment)\n",
    "# #         print(f.featureLocation)\n",
    "# #         print(f.modificationType.term[0])\n",
    "#         res['feature'].append(feature)\n",
    "  if hasattr(m, 'entityReference') and m.entityReference is not None:\n",
    "    e = m.entityReference\n",
    "    res['entityReference'] = {\n",
    "      'name': getattr(e, 'displayName', e.name),\n",
    "    }\n",
    "    if hasattr(e, 'xref') and e.xref is not None:\n",
    "      for x in e.xref:\n",
    "        if x.db is None:\n",
    "          continue\n",
    "        conv_id = '{0}:{1}'.format(convert_db_name(x.db), convert_db_id(x.id))\n",
    "#         if conv_id in all2mouse:\n",
    "#           for db_ref in all2mouse[conv_id]:\n",
    "#             if db_ref in gene_info_map:\n",
    "#               res['entityReference']['gene'] = gene_info_map[db_ref]\n",
    "        if 'gene' not in res['entityReference']:\n",
    "          if res['entityReference']['name'] in gene_name_map:\n",
    "            res['entityReference']['gene'] = gene_name_map[res['entityReference']['name']]\n",
    "          elif res['name'] in gene_name_map:\n",
    "            res['entityReference']['gene'] = gene_name_map[res['name']]\n",
    "        \n",
    "        # TODO: resolve small molecules\n",
    "        if ('uniprot' in x.db) \\\n",
    "          or ( \\\n",
    "            (x.db == 'ensembl' or x.db == 'ncbi gene') \\\n",
    "            and ( \\\n",
    "              'xref' not in res['entityReference'] \\\n",
    "              or res['entityReference']['xref']['db'] != 'ncbi gene' \\\n",
    "            ) \\\n",
    "          ) \\\n",
    "          or ( \\\n",
    "#             x.db in ['chebi'] \\\n",
    "#             and \n",
    "            'xref' not in res['entityReference'] \\\n",
    "          ):\n",
    "          # NCBI Gene has the top priority and the Ensemlb goes next\n",
    "          res['entityReference']['xref'] = {\n",
    "            'id': convert_db_id(x.id),\n",
    "            'db': convert_db_name(x.db),\n",
    "          }\n",
    "#       res['entityReference']['xref'] = [\n",
    "#         {\n",
    "#           'id': x.id,\n",
    "#           'db': x.db,\n",
    "# #           'relationshipType': [\n",
    "# #              r.term\n",
    "# #             for r in getattr(x, 'relationshipType', [])\n",
    "# #           ],\n",
    "#         }\n",
    "#         for x in e.xref\n",
    "#       ]\n",
    "#     if hasattr(e, 'organism') and e.organism is not None:\n",
    "#       o = e.organism\n",
    "#       res['entityReference']['organism'] = {\n",
    "#         'name': getattr(o, 'displayName', o.name),\n",
    "#       }\n",
    "#       if 'Homo' not in res['entityReference']['organism']['name']:\n",
    "#         print(res['entityReference']['organism']['name'])\n",
    "  return res\n",
    "\n",
    "\n",
    "def parse_reaction(r):\n",
    "  res = {\n",
    "    'type': 'reaction',\n",
    "    'name': r.displayName if r.displayName is not None else r.name,\n",
    "#     'name': getattr(r, 'displayName', r.name),\n",
    "#     'comment': getattr(r, 'comment', None),\n",
    "    'conversionDirection': getattr(r, 'conversionDirection', None),\n",
    "  }\n",
    "  if hasattr(r, 'left') and r.left is not None:\n",
    "    res['left'] = [parse_molecule(c) for _, c in enumerate(r.left)]\n",
    "  if hasattr(r, 'right') and r.right is not None:\n",
    "    res['right'] = [parse_molecule(c) for _, c in enumerate(r.right)]\n",
    "  return res\n",
    "\n",
    "\n",
    "def parse_control(c):\n",
    "  res = {\n",
    "    'type': 'control',\n",
    "    'name': c.displayName if c.displayName is not None else c.name,\n",
    "#     'name': getattr(c, 'displayName', c.name),\n",
    "#     'comment': getattr(c, 'comment', None),\n",
    "    'controlType': getattr(c, 'controlType', None),\n",
    "  }\n",
    "  if hasattr(c, 'controller') and c.controller is not None:\n",
    "    res['controller'] = [parse_molecule(cc) for cc in c.controller]\n",
    "  if hasattr(c, 'cofactor') and c.cofactor is not None:\n",
    "    res['cofactor'] = [parse_molecule(cc) for cc in c.cofactor]\n",
    "  if hasattr(c, 'controlled') and c.controlled is not None:\n",
    "    res['controlled'] = parse_reaction(c.controlled)\n",
    "  return res\n",
    "\n",
    "\n",
    "def parse_template_reaction(c):\n",
    "  res = {\n",
    "    'type': 'template_reaction',\n",
    "    'name': c.displayName if c.displayName is not None else c.name,\n",
    "#     'name': getattr(c, 'displayName', c.name),\n",
    "#     'comment': getattr(c, 'comment', None),\n",
    "    'templateDirection': getattr(c, 'templateDirection', None),\n",
    "  }\n",
    "  if hasattr(c, 'template') and c.template is not None:\n",
    "    res['template'] = parse_molecule(c.template)\n",
    "  if hasattr(c, 'product') and c.product is not None:\n",
    "    res['product'] = [parse_molecule(cc) for cc in c.product]\n",
    "  return res\n",
    "\n",
    "\n",
    "parsed_pathway_steps = {}\n",
    "def parse_pathway_step(s):\n",
    "  reactions = []\n",
    "  if s.name in parsed_pathway_steps:\n",
    "    # Avoid infinite recursions\n",
    "    return reactions\n",
    "  parsed_pathway_steps[s.name] = ''\n",
    "  res = {\n",
    "    'type': 'pathway_step',\n",
    "    'name': s.displayName if s.displayName is not None else s.name,\n",
    "#     'name': getattr(c, 'displayName', c.name),\n",
    "#     'comment': getattr(c, 'comment', None),\n",
    "    'stepDirection': getattr(s, 'stepDirection', None),\n",
    "  }\n",
    "  if hasattr(s, 'stepConversion') and s.stepConversion is not None:\n",
    "    res['stepConversion'] = parse_reaction(s.stepConversion)\n",
    "    reactions.append(res['stepConversion'])\n",
    "  if hasattr(s, 'stepProcess') and s.stepProcess is not None:\n",
    "    res['stepProcess'] = [parse_control(cc) for cc in s.stepProcess]\n",
    "    reactions.extend(res['stepProcess'])\n",
    "  if hasattr(s, 'nextStep') and s.nextStep is not None:\n",
    "    for ss in s.nextStep:\n",
    "      reactions.extend(parse_pathway_step(ss))\n",
    "  return reactions\n",
    "\n",
    "def parse_pathway(p):\n",
    "  res = {\n",
    "    'type': 'pathway',\n",
    "    'name': getattr(p, 'displayName', p.name),\n",
    "#     'comment': getattr(p, 'comment', None),\n",
    "  }\n",
    "  if hasattr(p, 'pathwayComponent') and p.pathwayComponent is not None:\n",
    "    res['pathwayComponent'] = []\n",
    "    for c in p.pathwayComponent:\n",
    "      if isinstance(c, Pathway):\n",
    "#         res_list.append(parse_pathway(c))\n",
    "        continue\n",
    "      elif isinstance(c, Control):\n",
    "        res['pathwayComponent'].append(parse_control(c))\n",
    "      elif isinstance(c, TemplateReaction):\n",
    "        res['pathwayComponent'].append(parse_template_reaction(c))\n",
    "      else:\n",
    "        res['pathwayComponent'].append(parse_reaction(c))\n",
    "  \n",
    "  if hasattr(p, 'pathwayOrder') and p.pathwayOrder is not None:\n",
    "    if 'pathwayComponent' not in res:\n",
    "      res['pathwayComponent'] = []\n",
    "    for s in p.pathwayOrder:\n",
    "      res['pathwayComponent'].extend(parse_pathway_step(s))\n",
    "\n",
    "  return res\n",
    "\n",
    "\n",
    "# print(list(pc.world.properties()))\n",
    "\n",
    "parsed_map = {}\n",
    "pathways = []\n",
    "\n",
    "generic_pathway_template = {\n",
    "  'type': 'pathway',\n",
    "  'name': '<generic>',\n",
    "  'pathwayComponent': [],\n",
    "}\n",
    "\n",
    "def process_reactions(reaction_class, parse_fn):\n",
    "  print(len(reaction_class.instances()))\n",
    "  for c in reaction_class.instances()[:]:\n",
    "    if c.name in parsed_map:\n",
    "      continue\n",
    "    parsed_map[c.name] = ''\n",
    "#     for xref in c.xref:\n",
    "#       print(xref.title)\n",
    "#       print(xref.get_properties())\n",
    "    r = parse_fn(c)\n",
    "    generic_pathway = copy.deepcopy(generic_pathway_template)\n",
    "    generic_pathway['pathwayComponent'].append(r)\n",
    "    pathways.append(generic_pathway) \n",
    "\n",
    "for reaction_class, parse_fn in [\n",
    "  (Control, parse_control),\n",
    "  (TemplateReaction, parse_template_reaction),\n",
    "  (BiochemicalReaction, parse_reaction),\n",
    "  (Conversion, parse_reaction),\n",
    "#   (Interaction, parse_reaction),\n",
    "]:\n",
    "  process_reactions(reaction_class, parse_fn)\n",
    "\n",
    "# for p in Pathway.instances()[:]:\n",
    "#   pathway = parse_pathway(p)\n",
    "# #   print(json.dumps(res, indent=2)[:100000])\n",
    "#   pathways.append(pathway)\n",
    "\n",
    "print(len(pathways))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "339\n",
      "[\n",
      "  \"R-HSA-158645\",\n",
      "  \"R-HSA-5336160\",\n",
      "  \"R-HSA-5607785\",\n",
      "  \"R-HSA-2648999\",\n",
      "  \"R-HSA-2730672\",\n",
      "  \"R-HSA-5358685\",\n",
      "  \"R-HSA-2684957\",\n",
      "  \"R-HSA-111972\",\n",
      "  \"R-HSA-3229230\",\n",
      "  \"R-HSA-5607784\",\n",
      "  \"R-HSA-5607787\",\n",
      "  \"R-HSA-5610119\",\n",
      "  \"R-HSA-8857690\",\n",
      "  \"R-HSA-5333036\",\n",
      "  \"Control_92e7377f8c17b64a542bbdb675a3fe12\",\n",
      "  \"R-HSA-1489501\",\n",
      "  \"R-HSA-1663719\",\n",
      "  \"R-HSA-9034483\",\n",
      "  \"R-HSA-5336148\",\n",
      "  \"R-HSA-9015052\",\n",
      "  \"Control_0f0070d86be5ad0b29e8311d985db98c\",\n",
      "  \"R-HSA-158643\",\n",
      "  \"R-HSA-977428\",\n",
      "  \"R-HSA-210503\",\n",
      "  \"Control_0224ec86f778e7febf2593ac5ddcc3b0\",\n",
      "  \"R-HSA-5333034\",\n",
      "  \"R-HSA-2855230\",\n",
      "  \"R-HSA-5211366\",\n",
      "  \"R-HSA-9034487\",\n",
      "  \"R-HSA-5626550\",\n",
      "  \"R-HSA-3223249\",\n",
      "  \"R-HSA-5607786\",\n",
      "  \"R-HSA-9034486\",\n",
      "  \"R-HSA-9034527\",\n",
      "  \"R-HSA-8876920\",\n",
      "  \"Control_656defa0440dd838b21dca1ca1196c61\",\n",
      "  \"R-HSA-71589\",\n",
      "  \"R-HSA-71542\",\n",
      "  \"R-HSA-2744259\",\n",
      "  \"R-HSA-6809098\",\n",
      "  \"R-HSA-8848729\",\n",
      "  \"R-HSA-3223247\",\n",
      "  \"R-HSA-5607091\",\n",
      "  \"R-HSA-2889054\",\n",
      "  \"R-HSA-9015053\",\n",
      "  \"Catalysis_33509413b4323e37244bb4583dcb65c9\",\n",
      "  \"Catalysis_a3f81fcffccdffe3bc33b1e092ff5af9\",\n",
      "  \"Catalysis_7481cae835c2e9d26d1f05f8dce63a4d\",\n",
      "  \"Catalysis_b8bd2880463202b778105e4a54a7c390\",\n",
      "  \"Catalysis_717745cedecdbcbabc71fba90cd9b352\",\n",
      "  \"Catalysis_48cd98212cd23ca1d052f05716cd8bbd\",\n",
      "  \"Catalysis_52d074e2708109ed2d72d5af8f0caba0\",\n",
      "  \"Catalysis_acb9217e6b0fd16f6dacd8af24923e3d\",\n",
      "  \"Catalysis_f17d3e17aaa9f4da57ed540c34048c87\",\n",
      "  \"Catalysis_6f3133eef0abe6b6cdaabc45eb5bc17d\",\n",
      "  \"Catalysis_910a33f75144aebc93215024e0394424\",\n",
      "  \"Catalysis_c5e184a642ef1673ce038fcaf278cba5\",\n",
      "  \"Catalysis_006a4404c144cd493090712faff1dac4\",\n",
      "  \"Catalysis_7ed7f71c5b59181addd9243e2c607d20\",\n",
      "  \"Catalysis_bd30e0601e9609171d8c2164cc514687\",\n",
      "  \"Catalysis_2011edeb5d01078f4211795062e7abf5\",\n",
      "  \"Catalysis_635fd1bf0d6a778e90df26d1bcef466e\",\n",
      "  \"Catalysis_6d38755eb412c0c8b0deebfecee03ae6\",\n",
      "  \"Catalysis_1cbd28e66a8f6cead8fa86f2b9c87f6b\",\n",
      "  \"Catalysis_f0989fcbfa9b52f844baab35334b2864\",\n",
      "  \"Catalysis_83a26e2011459cc136d07b3cc5fada26\",\n",
      "  \"Catalysis_fb7b3a6d2f7f582d3f7459e3fef2e62c\",\n",
      "  \"Catalysis_36b7df0ca3ddb2c0522385a138135cef\",\n",
      "  \"Catalysis_ba770f1d279683b55464723fa90b23f3\",\n",
      "  \"Catalysis_c057bfd7b3ca012212c4021192850933\",\n",
      "  \"Catalysis_915bd2bddb62f33af6c716cfe2a2da50\",\n",
      "  \"Catalysis_0bcb426f62204035f9a004843803cb04\",\n",
      "  \"Catalysis_e53aef662e271e791842408bd69ef301\",\n",
      "  \"Catalysis_fbb03456471b0d3f7def6a141db3dfe3\",\n",
      "  \"Catalysis_17560a6f06a9b07d5921c9dd913a7a18\",\n",
      "  \"Catalysis_b478e20709ab732143fefd18687d4a81\",\n",
      "  \"Catalysis_f2a2505bab5f407694e6f8a0f683f675\",\n",
      "  \"Catalysis_e341f6b1d3ddda5a083ba77429ee0760\",\n",
      "  \"Catalysis_e02489ac2806a416128b257ec0248ad1\",\n",
      "  \"Catalysis_1002d0b349d1db3a361727dd405ac74d\",\n",
      "  \"Catalysis_6d1b28361ac81229cdb48a6926622bff\",\n",
      "  \"Catalysis_63859491b32f693ad2bac3ef92227333\",\n",
      "  \"Catalysis_dcec3acda3c6c958349ce5d4887a7f31\",\n",
      "  \"Catalysis_ca4c2b90ced928f3342033ba2ded919f\",\n",
      "  \"Catalysis_2e668f8cc64372c032f5ebd6e16d32fd\",\n",
      "  \"Catalysis_3bd6c3e04760b674b8560e36518629e0\",\n",
      "  \"Catalysis_e436cc65222cbb1d0b1f76183d06084e\",\n",
      "  \"Catalysis_c6518a147bb17d032d20ffe427f6e7a8\",\n",
      "  \"Catalysis_e57d14355ec30b5660f14245e806759c\",\n",
      "  \"Catalysis_7468ee7b146986f9a26ace049a2dcaf7\",\n",
      "  \"Catalysis_c57ed00228d3424283730adeda4ec26e\",\n",
      "  \"Catalysis_7327506e8775fe06b777d8f52bf275cf\",\n",
      "  \"Catalysis_2ac170138cdaa9eee094f2ee7893bb41\",\n",
      "  \"Catalysis_7a7a702185363231e0b6a9974058d4ed\",\n",
      "  \"Catalysis_ed4914a3c48c7d7075f77856c5186c44\",\n",
      "  \"Catalysis_56d646996ebe5cd4ae879538c5c5badc\",\n",
      "  \"Catalysis_63e43081b7ef7419b999e89a8c53c0f3\",\n",
      "  \"Catalysis_c1d552c407575b892f2144ae974f077e\",\n",
      "  \"Catalysis_f3ff1807d287ad6d69738f5184df3b8d\",\n",
      "  \"Catalysis_adc9b289df04519db1f893c035aaa190\",\n",
      "  \"Catalysis_10dd73c0908f4fee7f88af3d621d46dd\",\n",
      "  \"Catalysis_125c5a9d3ac58e8a2fb55268c223669f\",\n",
      "  \"Catalysis_3f122a7c066498ad13489f795bb37d28\",\n",
      "  \"Catalysis_64e493d558049e4736b97af2e67132ce\",\n",
      "  \"Catalysis_0cb7806bfe4840b361ee5d6505e86dea\",\n",
      "  \"Catalysis_c2d38bfa951a078632369d3c12e7ab15\",\n",
      "  \"Catalysis_ad52fa185f218bc5e14e44e0947e6b0a\",\n",
      "  \"Catalysis_18a6a7b43af0583129fe07e1cb7b5d26\",\n",
      "  \"Catalysis_c795a9b209111948f911c2259a25309c\",\n",
      "  \"Catalysis_11519970180d7235e0584dc438ab133a\",\n",
      "  \"Catalysis_e9f7167590e735a38e93a4b9a06b33c5\",\n",
      "  \"Catalysis_d1bdc360c0e2af43d245560983dc8b90\",\n",
      "  \"Catalysis_8c3e2ed5c8838ae7ce699e17727e3e20\",\n",
      "  \"Catalysis_2cef99889ad4ad241f2abf1c32084548\",\n",
      "  \"Catalysis_6be46eeb1ecff283736b329e0970004e\",\n",
      "  \"Catalysis_b77933bae383327826536179f2d25d42\",\n",
      "  \"Catalysis_f856ecf184c2936842566f3e319159cf\",\n",
      "  \"Catalysis_903811ba9da4a3fcb2c080d2b1353473\",\n",
      "  \"Catalysis_afb0b0b21ee69aa90e341ada8a20a73b\",\n",
      "  \"Catalysis_0a92f9fc7f1f965dca4699434c3346d9\",\n",
      "  \"Catalysis_b1c8f06a634948a4793ebfc43cc547f5\",\n",
      "  \"Catalysis_61ec17c7e9d636a6e70ed622c1d8641f\",\n",
      "  \"Catalysis_efbf5ec5edad5c0611364d8b2521a7a7\",\n",
      "  \"Catalysis_88964d26a5ccaf8b5bf54993e074d46c\",\n",
      "  \"Catalysis_abcb89108bb1be3bcc6966a069dbb8ac\",\n",
      "  \"Catalysis_2131384ae82c61e79d5bb261cc30e1c6\",\n",
      "  \"Catalysis_dd8187a3095b30b282b50ef27f2178ff\",\n",
      "  \"Catalysis_5ac9e145bf9bfddab11ccc3932c30c61\",\n",
      "  \"Catalysis_7455600125aaba997296e6051270c31a\",\n",
      "  \"Catalysis_b7dfe873db3d6b4b470ca7f65c6493cb\",\n",
      "  \"Catalysis_5e05e18945598befa684d8747d7c7510\",\n",
      "  \"Catalysis_97ef44a47d40595c3a2a625a23c448cb\",\n",
      "  \"Catalysis_51dbf04f405ddb1c0e752b713ccc9fb6\",\n",
      "  \"Catalysis_e24b6e60e1e375de805a38cf7423673d\",\n",
      "  \"Catalysis_31608980d1cc5e71fc37bfe4e5425b03\",\n",
      "  \"Catalysis_1f3d2b1063ac7f1e8d9a771cfc644ae2\",\n",
      "  \"Catalysis_895247441a612ba119b019eb4c6edee3\",\n",
      "  \"Catalysis_f00fb05179a9be205824c25eb64a7ba3\",\n",
      "  \"Catalysis_0f45c9af5c7ddbb6f9ffbc9383d379df\",\n",
      "  \"Catalysis_56a17e76c27e50a9a1d7d8d9ea7557ad\",\n",
      "  \"Catalysis_c3999f288f716e467bedd50ab129cdbc\",\n",
      "  \"Catalysis_c2182e9d70b6a4ae128bce87b97741ff\",\n",
      "  \"Catalysis_f4b90b40e2a4cd3163c0b657e2acdd64\",\n",
      "  \"Calcium binds calmodulin at the synapse\",\n",
      "  \"PPM1F dephosphorylates p-T286-CaMKII\",\n",
      "  \"Exocytosis of Insulin\",\n",
      "  \"E-cadherin degradation by ADAM10, ADAM15\",\n",
      "  \"factor V -> factor Va + factor V activation peptide\",\n",
      "  \"Activation of Ca permeable AMPA receptors\",\n",
      "  \"Decreasing cGMP concentration promotes intracellular Ca2+ release in response to WNT\",\n",
      "  \"CaMKII autophosphorylates\",\n",
      "  \"Activated calmodulin binds ADCY1,ADCY8\",\n",
      "  \"TPCN1/2 transport lysosomal Ca2+ to cytosol\",\n",
      "  \"CALM1:4xCa2+ binds GluN1:GluN2 (GRIN1:GRIN2) NMDA receptors and displaces ACTN2\",\n",
      "  \"PPM1F dephosphorylates CAMK1\",\n",
      "  \"PPM1E dephosphorylates CAMK4\",\n",
      "  \"SPARC binds Collagen type I fibril, hydroxylapatite and Ca2+\",\n",
      "  \"thrombin:cleaved antithrombin III:heparin -> thrombin:cleaved antithrombin III + heparin\",\n",
      "  \"Dissociation of CaM and CAMK2 autophosphorylation\",\n",
      "  \"G-protein subunits dissociate from WNT:FZD complex\",\n",
      "  \"WNT5A and WNT11 bind FZD receptors to initiate Ca2+ signaling\",\n",
      "  \"WNT:FZD complex promotes G-protein nucleotide exchange\",\n",
      "  \"FZD recruits trimeric G-proteins\",\n",
      "  \"Autophosphorylation and activation of CAMK2\",\n",
      "  \"Translocation of NFATC1:CaN:CaM to nucleus\",\n",
      "  \"Active calmodulin binds CAMK2\",\n",
      "  \"Translocation and activation of PKC alpha in response to WNT signaling\",\n",
      "  \"Calcineurin binds and dephosphorylates NFAT1 in response to WNT/Ca2+ signaling\",\n",
      "  \"Activation of Calcineurin\",\n",
      "  \"SERPINA5 binds activated protein C\",\n",
      "  \"S100A8:S100A9:AA:Ca(2+) binds NOX2 complex\",\n",
      "  \"S100A8:S100A9 binds Mn2+\",\n",
      "  \"Activated calmodulin dissociates from CaMKII-gamma\",\n",
      "  \"Calcium Influx through Voltage-gated Calcium Channels\",\n",
      "  \"Activation of Ca-permeable Kainate receptors\",\n",
      "  \"CAMK4 autophosphorylates\",\n",
      "  \"Protease nexin-1 (PN1) is a specific and extremely efficient inhibitor of thrombin.\",\n",
      "  \"Insulin secretory granule translocates to cell cortex\",\n",
      "  \"Proinsulin:Zinc:Calcium complex translocates to immature secretory granule\",\n",
      "  \"SLC24A1 exchanges 4Na+ for Ca2+, K+\",\n",
      "  \"Proinsulin binds zinc and calcium forming Proinsulin:zinc:calcium\",\n",
      "  \"Calcium-sensing receptor binds calcium\",\n",
      "  \"CaMKKs phosphorylate CAMK1\",\n",
      "  \"Ca2+ influx through voltage gated Ca2+ channels\",\n",
      "  \"CaMKII binds activated calmodulin\",\n",
      "  \"CaMKK autophosphorylates in the nucleus\",\n",
      "  \"p-T287-CAMK2G:CALM1:4xCa2+ enters nucleus\",\n",
      "  \"Calmodulin binds CAMK4\",\n",
      "  \"CaMKK phosphorylates CAMK4\",\n",
      "  \"CaMKK binds activated calmodulin in the nucleus\",\n",
      "  \"Activation of RasGRF\",\n",
      "  \"PA63:ANTRX1 oligomer transports EF and LF (target cell endosome to cytosol)\",\n",
      "  \"Endocytosis of EF:LF:PA63:ANTXR2 (plasma membrane to endosome membrane)\",\n",
      "  \"ANTXR2-bound PA63 forms oligomers\",\n",
      "  \"PA83 binds ANTXR1\",\n",
      "  \"PA83 binds ANTXR2\",\n",
      "  \"Furin cleaves ANTXR2-bound PA83 to yield PA63\",\n",
      "  \"ANTXR1-bound PA63 forms oligomers\",\n",
      "  \"EF and LF bind to PA63:ANTXR2 oligomer\",\n",
      "  \"EF and LF bind to PA63:ANTXR1 oligomer\",\n",
      "  \"Furin cleaves ANTXR1-bound PA83 to yield PA63\",\n",
      "  \"PA63:ANTRX2 oligomer transports EF and LF (target cell endosome to cytosol)\",\n",
      "  \"Endocytosis of EF:LF:PA63:ANTXR1 (plasma membrane to endosome membrane)\",\n",
      "  \"CL-LK binds carbohydrates on target cell surface\",\n",
      "  \"SLC8A3 (NCX3) exchanges sodium (cytosol) for calcium (mitochondrial intermembrane space)\",\n",
      "  \"Insulin secretory granule translocates across the cortical actin network\",\n",
      "  \"Carboxypeptidase E cleaves Insulin(25-56) to yield Insulin(25-54)\",\n",
      "  \"PCSK1 cleaves proinsulin to yield Insulin(25-56) and Insulin(57-110)\",\n",
      "  \"PCSK2 cleaves Insulin(57-110) to yield Insulin(90-110) and C-peptide (Insulin(57-89))\",\n",
      "  \"Insulin secretory granule docks at the plasma membrane\",\n",
      "  \"Transport of Ca++ from platelet dense tubular system to cytoplasm\",\n",
      "  \"Binding of IP3 to IP3 receptor\",\n",
      "  \"Calcium and Diacylglycerol activate CalDAG-GEFs (RasGRPs)\",\n",
      "  \"Hyaluronidase 2 (HYAL2) hydrolyses HA into 20kDa fragments\",\n",
      "  \"S100A8:S100A9 binds Zn2+\",\n",
      "  \"S100A7/S100A15 binds Zn2+\",\n",
      "  \"GRK2 binds CALM1:4Ca2+\",\n",
      "  \"SERPIND1 binds activated thrombin (factor IIa)\",\n",
      "  \"CALM1:Ca2+ binds CNG channel\",\n",
      "  \"Calcium influx via CRAC\",\n",
      "  \"activation of factor VII\",\n",
      "  \"VDAC1,2,3 translocate calcium from the cytosol to the mitochondrial intermembrane space\",\n",
      "  \"p-AKT2 phosphorylates C2CD5\",\n",
      "  \"prothrombin -> activated thrombin (factor IIa) + thrombin activation peptide (Xa catalyst)\",\n",
      "  \"activation of factor X by TF:F7a\",\n",
      "  \"Protein Kinase C beta (PKC-beta, PRKCB) is activated\",\n",
      "  \"LTCC heteropentamer (open) transports Ca2+ from extracellular region to cytosol\",\n",
      "  \"Activation of conventional Protein Kinase C\",\n",
      "  \"protein C -> activated protein C + protein C heavy chain activation peptide\",\n",
      "  \"Phospho-cPLA2 translocates to membranes when intracellular calcium levels increase\",\n",
      "  \"Calcium binds calmodulin\",\n",
      "  \"MCU translocates calcium from the mitochondrial intermembrane space to the mitochondrial matrix\",\n",
      "  \"SLC8B1 (NCLX) exchanges sodium (mitochondrial intermembrane space) for calcium (mitochondrial matrix)\",\n",
      "  \"LETM1 exchanges protons (mitochondrial intermembrane space) for calcium (mitochondrial matrix)\",\n",
      "  \"Ficolin-3 binds to molecular patterns on the target cell surface\",\n",
      "  \"activation of factor X by VIIIa:IXa\",\n",
      "  \"STIM1 oligomerizes\",\n",
      "  \"TRPs transport extracellular Ca2+ to cytosol\",\n",
      "  \"Translocation of CaN:CaM:NFAT to nucleus\",\n",
      "  \"prothrombin -> activated thrombin (factor IIa) + thrombin activation peptide (prothrombinase catalyst)\",\n",
      "  \"activation of factor X by TF:F7\",\n",
      "  \"formation of TF:F7 complex\",\n",
      "  \"formation of TF:F7a complex\",\n",
      "  \"Inactivation of factor VIIa by TFPI\",\n",
      "  \"Calcium Binds Caldesmon\",\n",
      "  \"Myosin Binds ATP\",\n",
      "  \"ATP Hydrolysis By Myosin\",\n",
      "  \"MYLK (MLCK) Active Calmodulin Binding\",\n",
      "  \"Release Of ADP From Myosin\",\n",
      "  \"Calcineurin dephosphorylates NFATC1,2,3\",\n",
      "  \"Calcineurin binds NFATC1,2,3\",\n",
      "  \"SLC24A5 exchanges cytosolic 4Na+ for Golgi luminal Ca2+, K+\",\n",
      "  \"DAG and Ca+2 bind to PKC and tether it to membrane\",\n",
      "  \"Activated protein C binds Protein S\",\n",
      "  \"RYR tetramers transport Ca2+ from sarcoplasmic reticulum lumen to cytosol\",\n",
      "  \"RCVRN:Ca2+ binds to and inhibits GRK1\",\n",
      "  \"CALB1 binds 4xCa2+\",\n",
      "  \"cGMP:CNG transports Na+ and Ca2+ into the rod outer segment\",\n",
      "  \"Factor IIa inhibitors (peptide) binds IIa\",\n",
      "  \"Calcium Influx via TRPC1\",\n",
      "  \"factor Va + factor Xa -> Va:Xa complex (prothrombinase)\",\n",
      "  \"SLC24A1-4 exchange extracellular 4Na+ for cytosolic Ca2+, K+\",\n",
      "  \"ATP2C1/2:Mg2+ transport cytosolic Ca2+ to Golgi lumen\",\n",
      "  \"Connection of adjacent cells through calcium-dependent trans-dimerization of cadherin\",\n",
      "  \"E-cadherin degradation by MMP9, KLK7\",\n",
      "  \"E-cadherin degradation by PS1:NCSTN (Gamma-secretase)\",\n",
      "  \"AKT1 phosphorylates eNOS\",\n",
      "  \"The cofactor BH4 is required for electron transfer in the eNOS catalytic cycle\",\n",
      "  \"eNOS:Caveolin-1 complex binds to CaM\",\n",
      "  \"HSP90 binds eNOS:Caveolin-1:CaM complex\",\n",
      "  \"AKT1 binds eNOS complex via HSP90\",\n",
      "  \"Caveolin-1 dissociates from eNOS:CaM:HSP90 complex\",\n",
      "  \"IP3R:I(1,4,5)P3 tetramer transports Ca2+ from ER lumen to cytosol\",\n",
      "  \"Influx of extracellular calcium\",\n",
      "  \"Calpain activation\",\n",
      "  \"BH2 binding can lead to eNOS uncoupling\",\n",
      "  \"K+-independent Li+/Ca2+ exchanger transport\",\n",
      "  \"SLC8A1,2,3 exchange 3Na+ for Ca2+\",\n",
      "  \"Defective SLC24A5 does not exchange cytosolic 4Na+ for Golgi luminal Ca2+, K+\",\n",
      "  \"PMCA extrusion of Ca2+\",\n",
      "  \"CaMKII binds activated NMDA receptor\",\n",
      "  \"thrombin:antithrombin III:heparin -> thrombin:cleaved antithrombin III:heparin\",\n",
      "  \"GUCA binds Ca2+ to form GUCA:Ca2+\",\n",
      "  \"Proteinase-activated receptors can bind thrombin\",\n",
      "  \"S100A1 binds TLR4:LY96\",\n",
      "  \"P2X1-mediated entry of Ca++ from plasma\",\n",
      "  \"Sequestration of Ca2+ to dense tubular network lumen\",\n",
      "  \"Myosin Binds ATP\",\n",
      "  \"Inactive catalytic PP2B is activated by the binding of calmodulin\",\n",
      "  \"ATP Hydrolysis By Myosin\",\n",
      "  \"Active ARSA translocates to the lysosome\",\n",
      "  \"Thrombin binding to GP1b:IX:V\",\n",
      "  \"Defective SLC24A1 does not exchange extracellular 4Na+ for cytosolic Ca2+, K+\",\n",
      "  \"Defective SLC24A4 does not exchange extracellular 4Na+ for cytosolic Ca2+, K+\",\n",
      "  \"INTL1 binds bacterial glycans\",\n",
      "  \"HA:HAR:HYAL2 binds to SLC9A1:p-CHP:Ca2+\",\n",
      "  \"Serum amyloid P-component forms homopentamers\",\n",
      "  \"E-cadherin degradation by caspase-3 and calpain-1\",\n",
      "  \"CAMK1 binds calmodulin\",\n",
      "  \"Activated protein C is released from PROCR\",\n",
      "  \"IQGAP1 binds CDH1:CTTNB1:CTTNA1 and MEN1\",\n",
      "  \"IQGAPs bind CALM1\",\n",
      "  \"Calcineurin binds and dephosphorylates NFAT\",\n",
      "  \"Factor Xa inhibitors bind Va:Xa\",\n",
      "  \"Ca2+ influx into the post-synaptic cell\",\n",
      "  \"Activated protein C:PROCR binds F2R\",\n",
      "  \"MBL binds to repetitive carbohydrate structures on the surfaces of viruses, bacteria, fungi, and protozoa\",\n",
      "  \"E-cadherin degradation by MMP3, MMP7 and plasmin.\",\n",
      "  \"Calcium Binds Troponin-C\",\n",
      "  \"Soluble PROCR binds activated protein C\",\n",
      "  \"RasGRP interacts with DAG and Ca2+\",\n",
      "  \"HTR3 pentamer:5HT transports Na+,K+, and Ca2+\",\n",
      "  \"CatSper Channel Mediated Calcium Transport\",\n",
      "  \"Activation of highly calcium permeable nicotinic acetylcholine receptors\",\n",
      "  \"Factor IIa inhibitors (compounds) binds IIa\",\n",
      "  \"CD209 binds ICAM3\",\n",
      "  \"activated thrombin (factor IIa) + thrombomodulin -> activated thrombin:thrombomodulin\",\n",
      "  \"Activation of highly calcium permeable postsynaptic nicotinic acetylcholine receptors\",\n",
      "  \"CBLL1 binds SRC-phosphorylated CDH1 complex\",\n",
      "  \"Sequestration of Ca2+ to intracellular stores\",\n",
      "  \"CBLL1 ubiqutinates the InlA-bound CDH1 complex\",\n",
      "  \"CACFD1 dimer transports Ca2+ from extracellular region to cytosol\",\n",
      "  \"PLA2G2A binds bacterial phospholipids\",\n",
      "  \"SRC phosphorylates InlA-bound CDH1 and CTNNB1\",\n",
      "  \"Ficolin-1 binds to molecular patterns on the target cell surface\",\n",
      "  \"Release Of ADP From Myosin\",\n",
      "  \"Ficolin-2 binds to molecular patterns on the target cell surface\",\n",
      "  \"ALOX5 is phosphorylated by MAPKAP2\",\n",
      "  \"CaMKK autophosphorylates in the cytosol\",\n",
      "  \"InlA binds CDH1\",\n",
      "  \"CaMKK binds activated calmodulin in the cytosol\",\n",
      "  \"Formation of serum amyloid P decamer\",\n",
      "  \"Serum amyloid P binds DNA and chromatin\",\n",
      "  \"Amyloid fibrils have additional components\",\n",
      "  \"factor Va -> factor Vi\",\n",
      "  \"Factor Xa inhibitors binds Xa\",\n",
      "  \"DAG and Ca+2 bind to PKC and tether it to membrane\",\n",
      "  \"factor XIII cleaved tetramer + 2 Ca++ -> factor XIIIa + 2 factor XIII B chain\",\n",
      "  \"activated thrombin (factor IIa) + antithrombin III:heparin -> thrombin:antithrombin III:heparin\"\n",
      "]\n"
     ]
    }
   ],
   "source": [
    "def match_participant(m, query):\n",
    "  if m['type'] == 'complex':\n",
    "    return any([match_participant(c, query) for c in m['component']])\n",
    "  if query.lower() in m['name'].lower():\n",
    "    return True\n",
    "  return False\n",
    "\n",
    "def filter_pathways_by_participant(query):\n",
    "  filtered_pathways = []\n",
    "  for p in pathways:\n",
    "    filtered_p = copy.deepcopy(p)\n",
    "    filtered_p['pathwayComponent'] = []\n",
    "    for r in p['pathwayComponent']:\n",
    "      matched = False\n",
    "      rr = r\n",
    "      if r['type'] == 'control':\n",
    "        for m in r['controller']:\n",
    "          matched = matched or match_participant(m, query)\n",
    "        for m in r['cofactor']:\n",
    "          matched = matched or match_participant(m, query)\n",
    "      elif r['type'] == 'template_reaction':\n",
    "        for m in r['product']:\n",
    "          matched = matched or match_participant(m, query)\n",
    "      if 'left' in rr:\n",
    "        for m in rr['left']:\n",
    "          matched = matched or match_participant(m, query)\n",
    "      if 'right' in rr:\n",
    "        for m in rr['right']:\n",
    "          matched = matched or match_participant(m, query)\n",
    "    if matched:\n",
    "      filtered_p['pathwayComponent'].append(r)\n",
    "    if len(filtered_p['pathwayComponent']) > 0:\n",
    "      filtered_pathways.append(filtered_p)\n",
    "  return filtered_pathways\n",
    "\n",
    "filtered_pathways = filter_pathways_by_participant('ca2+')\n",
    "print(len(filtered_pathways))\n",
    "print(json.dumps(list(map(lambda p: p['pathwayComponent'][0]['name'], filtered_pathways)), indent=2))\n",
    "#print(json.dumps(list(map(lambda p: p['pathwayComponent'][0]['controlled']['name'], filtered_pathways)), indent=2))\n",
    "#print(json.dumps(filtered_pathways[:5], indent=2))\n",
    "\n",
    "# https://reactome.org/PathwayBrowser/#/R-HSA-163200&SEL=R-HSA-163213&PATH=R-HSA-1430728,R-HSA-1428517\n",
    "# https://reactome.org/PathwayBrowser/#/R-HSA-163200&SEL=R-HSA-163214&PATH=R-HSA-1430728,R-HSA-1428517\n",
    "\n",
    "# for p in pathways:\n",
    "#   for r in p['pathwayComponent']:\n",
    "#     print(r['name'])\n",
    "#     break\n",
    "#     if 'R11945' in r['name']:\n",
    "#       print(r)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio.Data.CodonTable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TTT': 'F',\n",
       " 'TTC': 'F',\n",
       " 'TTA': 'L',\n",
       " 'TTG': 'L',\n",
       " 'TCT': 'S',\n",
       " 'TCC': 'S',\n",
       " 'TCA': 'S',\n",
       " 'TCG': 'S',\n",
       " 'TAT': 'Y',\n",
       " 'TAC': 'Y',\n",
       " 'TGT': 'C',\n",
       " 'TGC': 'C',\n",
       " 'TGG': 'W',\n",
       " 'CTT': 'L',\n",
       " 'CTC': 'L',\n",
       " 'CTA': 'L',\n",
       " 'CTG': 'L',\n",
       " 'CCT': 'P',\n",
       " 'CCC': 'P',\n",
       " 'CCA': 'P',\n",
       " 'CCG': 'P',\n",
       " 'CAT': 'H',\n",
       " 'CAC': 'H',\n",
       " 'CAA': 'Q',\n",
       " 'CAG': 'Q',\n",
       " 'CGT': 'R',\n",
       " 'CGC': 'R',\n",
       " 'CGA': 'R',\n",
       " 'CGG': 'R',\n",
       " 'ATT': 'I',\n",
       " 'ATC': 'I',\n",
       " 'ATA': 'I',\n",
       " 'ATG': 'M',\n",
       " 'ACT': 'T',\n",
       " 'ACC': 'T',\n",
       " 'ACA': 'T',\n",
       " 'ACG': 'T',\n",
       " 'AAT': 'N',\n",
       " 'AAC': 'N',\n",
       " 'AAA': 'K',\n",
       " 'AAG': 'K',\n",
       " 'AGT': 'S',\n",
       " 'AGC': 'S',\n",
       " 'AGA': 'R',\n",
       " 'AGG': 'R',\n",
       " 'GTT': 'V',\n",
       " 'GTC': 'V',\n",
       " 'GTA': 'V',\n",
       " 'GTG': 'V',\n",
       " 'GCT': 'A',\n",
       " 'GCC': 'A',\n",
       " 'GCA': 'A',\n",
       " 'GCG': 'A',\n",
       " 'GAT': 'D',\n",
       " 'GAC': 'D',\n",
       " 'GAA': 'E',\n",
       " 'GAG': 'E',\n",
       " 'GGT': 'G',\n",
       " 'GGC': 'G',\n",
       " 'GGA': 'G',\n",
       " 'GGG': 'G'}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(Bio.Data.CodonTable.standard_dna_table).forward_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "mol_data = []\n",
    "type_id = 0\n",
    "\n",
    "mol_data.extend([\n",
    "  {\n",
    "    'type': type_id + 0,\n",
    "    'category': 'lipid',\n",
    "    'name': 'Generic lipid',\n",
    "    'radius': 0.0025,\n",
    "  },\n",
    "  {\n",
    "    'type': type_id + 1,\n",
    "    'category': 'protein',\n",
    "    'name': 'Generic protein',\n",
    "    'radius': 0.0025,\n",
    "  }\n",
    "])\n",
    "\n",
    "type_id = 100\n",
    "# Generic RNA codons\n",
    "mol_data.extend([\n",
    "  {\n",
    "    'type': type_id + i,\n",
    "    'category': 'rna',\n",
    "    'name': codon,\n",
    "    'radius': 0.00075,\n",
    "  }\n",
    "  for i, (codon, amino_acid) in enumerate(Bio.Data.CodonTable.standard_dna_table.forward_table.items())\n",
    "])\n",
    "type_id += len(Bio.Data.CodonTable.standard_dna_table.forward_table)\n",
    "\n",
    "# Start/stop RNA codons\n",
    "mol_data.extend([\n",
    "  {\n",
    "    'type': type_id + i,\n",
    "    'category': 'rna',\n",
    "    'name': codon,\n",
    "    'radius': 0.00075,\n",
    "  }\n",
    "  for i, codon in enumerate(Bio.Data.CodonTable.standard_dna_table.start_codons + Bio.Data.CodonTable.standard_dna_table.stop_codons)\n",
    "])\n",
    "\n",
    "# Write to a file\n",
    "with open('./gen/large-molecules.json', 'wt') as f:\n",
    "  f.write(json.dumps(mol_data, indent=2))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
